### DISCUSSION TITLE
How to use Jinja2 Template with Nornir 3

### DISCUSSION THREAD

**Discussion by nouse4it** [nornir-automation/nornir#664] [Q&A] [answered]
I'm seeing a lot of example on the web, using Jinja2 Template with nornir, by using the `task=template_file` to create a template as host attribute.
But it´s seems like this functions doesn´t exists in Nornir 3 anymore?
Can someone explain me how can I configure a Device with data from a template?

What I´m trying to do is to read all interfaces from a device and load an additional vars file with host specific informations of the device. 
Then pass the interfaces to a function which creates a config file based on a Jinja2 Template.
There are two Jinja2 Templates. One for configuring 802.1x per interface and one to use Port-Security on the interfaces.
The script should choose the Template based on the additional host info from the var file.
After that the script should create a config file based on the process above and configure the interfaces.

My code looks like this:

```
# Importing all needed Modules
from nornir import InitNornir
from nornir.core.task import Task, Result
from nornir_utils.plugins.functions import print_result
from nornir_napalm.plugins.tasks import napalm_get
from nornir_netmiko import netmiko_send_command, netmiko_send_config
from nornir.core.filter import F
from nornir_utils.plugins.tasks.data import load_yaml
import getpass
import jinja2

nr = InitNornir(config_file="config.yaml")

#==============================================================================
# Get Interfaces to run Config against
def get_intf(task):
    r = task.run(task=netmiko_send_command, command_string="show int status", use_genie=True)

#==============================================================================
# Parse Results from Nornir Task Result
def result_parser(data):
    for host, task_result in data.items():
        return task_result[1].result

#==============================================================================
# Load and Render J2 Template for Interfaces. 
def config_intf(task):
    loader = jinja2.FileSystemLoader('templates')
    jenv = jinja2.Environment(loader=loader, trim_blocks=True, lstrip_blocks=True)
    vars_yaml = f"{task.host}.yaml"
    vars_data = task.run(task=load_yaml, file=vars_yaml)
    if vars_data.result['NAC']['dot1x']['enabled'] == 'yes':
        j2template = vars_data.result['NAC']['dot1x']['template']
    else:
        j2template = 'port_security_it_is_then'
    template = jenv.get_or_select_template(j2template)
    parsed_template = template.render(interfaces)
    print(parsed_template)
    r = task.run(task=template_file, template=parsed_template, path="")
    task.host['intf_config'] = r.result
    task.run(task=netmiko_send_config, name="Configuring Interfaces", config_command=task.host["intf_config"])

intf_getter = nr.run(task=get_intf) # get interfaces from device
interfaces = result_parser(intf_getter) # create dict of interfaces from nornir result (intf_getter)
result = nr.run(task=config_intf) # create config based on template and configure device
print_result(result)
```

When I run the script, i kept getting told that the name 'template_file' is not definied
Can someone help me on what I´m missing here?

Thank you!

**Comment by lucasmarcel** [comment, marked as answer]
With Nornir 3, you need to install the nornir-jinja2 plugin:
`pip install nornir-jinja2`. Then in your script:
`from nornir_jinja2.plugins.tasks import template_file`

The documentation is here: [https://nornir.tech/nornir_jinja2/html/](https://nornir.tech/nornir_jinja2/html/)
